core/src/Cabana_CommunicationPlan.hpp

```diff
@@ -0,0 +1,602 @@
+{
+  public:
+
+    // Cabana memory space.
+    using memory_space = MemorySpace;
+
+    // Kokkos execution space.
+    using execution_space = typename memory_space::execution_space;
+
+    /*!
+      \brief Constructor.
+
+      \param comm The MPI communicator over which the distributor is defined.
+    */
+    CommunicationPlan( MPI_Comm comm )
+        : _comm( comm )
+    {}
+
+    /*!
+      \brief Get the MPI communicator.
+    */
+    MPI_Comm comm() const
+    { return _comm; }
+
+    /*!
+      \brief Get the number of neighbor ranks that this rank will communicate
+      with.
+
+      \return The number of MPI ranks that will exchange data with this rank.
+    */
+    int numNeighbor() const
+    { return _neighbors.size(); }
+
+    /*!
+      \brief Given a local neighbor id get its rank in the MPI communicator.
+
+      \param neighbor The local id of the neighbor to get the rank for.
+
+      \return The MPI rank of the neighbor with the given local id.
+    */
+    int neighborRank( const int neighbor ) const
+    { return _neighbors[neighbor]; }
+
+    /*!
+      \brief Get the number of elements this rank will export to a given neighbor.
+
+      \param neighbor The local id of the neighbor to get the number of
+      exports for.
+
+      \return The number of elements this rank will export to the neighbor with the
+      given local id.
+     */
+    std::size_t numExport( const int neighbor ) const
+    { return _num_export[neighbor]; }
+
+    /*!
+      \brief Get the total number of exports this rank will do.
+
+      \return The total number of elements this rank will export to its
+      neighbors.
+    */
+    std::size_t totalNumExport() const
+    { return _total_num_export; }
+
+    /*!
+      \brief Get the number of elements this rank will import from a given neighbor.
+
+      \param neighbor The local id of the neighbor to get the number of
+      imports for.
+
+      \return The number of elements this rank will import from the neighbor
+      with the given local id.
+     */
+    std::size_t numImport( const int neighbor ) const
+    { return _num_import[neighbor]; }
+
+    /*!
+      \brief Get the total number of imports this rank will do.
+
+      \return The total number of elements this rank will import from its
+      neighhbors.
+    */
+    std::size_t totalNumImport() const
+    { return _total_num_import; }
+
+    /*!
+      \brief Get the number of export elements.
+
+      Whenever the communciation plan is applied, this is the total number of
+      elements expected to be input on the sending ranks (in the forward
+      communication plan). This will be different than the number returned by
+      totalNumExport() if some of the export ranks used in the construction
+      are -1 and therefore will not particpate in an export operation.
+
+      \return The number of export elements.
+    */
+    std::size_t numExportElement() const
+    { return _num_export_element; }
```

Comment:
I understand we need something like this to validate the user input when executing the communication plan but I don't like the name of this member function because I don't find it very descriptive.

Reply:
I have changed the name as we discussed